Topics/System_Design/behavioral_design_patterns.md

```diff
@@ -0,0 +1,120 @@
+
+1. **Context:** You, as the traveler, represent the context in this scenario. You have a goal (getting to your destination) but are open to different strategies for achieving it.
+
+2. **Strategy Interface:** This is an abstract representation of the transportation strategy. It defines a common interface for all transportation methods, such as a TravelStrategy interface with a method travel(). This interface ensures that all strategies can be used interchangeably.
+
+3. **Concrete Strategies:** These are the different transportation methods - car, train, bus, and airplane. Each concrete strategy implements the TravelStrategy interface and provides its own implementation of the travel() method. For example, the CarTravelStrategy might involve driving your own car or renting one, the TrainTravelStrategy might involve booking a train ticket, and so on.
+
+4. **Context with Strategy:** You, as the traveler, hold a reference to a specific transportation strategy. Depending on your requirements, preferences, and circumstances, you can dynamically switch between different strategies without affecting the rest of your vacation plans.
+
+In this example, the Strategy pattern allows you to decouple the selection of transportation method from the vacation planning process. You can easily switch between different strategies based on factors like distance, cost, and time constraints without having to modify the rest of your vacation itinerary. For instance, if you initially planned to travel by train but find that flights are more affordable, you can switch to the airplane strategy without rewriting your entire travel plan.
+
+Overall, the Strategy pattern promotes flexibility, modularity, and maintainability by encapsulating algorithms or strategies in separate classes and allowing them to be selected or swapped at runtime based on specific requirements or conditions.
+
+## Iterator Design Pattern <a name="itr"></a>
+
+The Iterator design pattern is a behavioral pattern that provides a way to access elements of an aggregate object sequentially without exposing its underlying representation. It allows clients to traverse the elements of a collection without needing to know its internal structure. The Iterator pattern typically consists of two main components: the Iterator interface or class, which defines methods for accessing elements sequentially, and the Aggregate interface or class, which defines a method for creating an iterator object. By decoupling the traversal algorithm from the collection, the Iterator pattern enhances the flexibility and reusability of both the collection and the traversal algorithm. It's commonly used in scenarios where collections vary in structure or where there's a need to iterate over elements in a uniform manner.
+
+<p align="center">
+<img width="709" alt="Screenshot 2024-03-17 at 9 10 00 AM" src="https://github.com/paridhi26/CSC311Project/assets/85652265/48ac92bc-f2eb-4c04-993a-532ca2cf1b76">
```

Comment:
The image is not getting generated for me.

Reply:
should be fixed!